File: linkgame-client/src/main/java/io/github/octcarp/linkgame/client/net/LobbyData.java
package io.github.octcarp.linkgame.client.net;

import io.github.octcarp.linkgame.client.controller.LobbyController;
import io.github.octcarp.linkgame.client.utils.SceneSwitcher;
import io.github.octcarp.linkgame.common.packet.Request;
import io.github.octcarp.linkgame.common.packet.RequestType;
import javafx.application.Platform;

import java.util.List;

public class LobbyData {
    private static final LobbyData instance = new LobbyData();

    private List<String> waitingPlayers;
    private String lastOpponent;

    private boolean isWaiting = false;

    private LobbyData() {
    }

    public static LobbyData getInstance() {
        return instance;
    }

    public void enterLobby() {
        Request request = new Request(RequestType.ENTER_LOBBY);
        ClientService.getInstance().sendRequest(request);
    }

    public void exitLobby() {
        Request request = new Request(RequestType.EXIT_LOBBY);
        ClientService.getInstance().sendRequest(request);
    }

    public void startWaiting() {
        Request request = new Request(RequestType.START_WAITING);
        ClientService.getInstance().sendRequest(request);
    }

    public void stopWaiting() {
        Request request = new Request(RequestType.STOP_WAITING);
        ClientService.getInstance().sendRequest(request);
    }

    public void reAllWaitingPlayers(List<String> players) {
        Platform.runLater(() -> {
            LobbyController controller = (LobbyController)
                    SceneSwitcher.getInstance().getController("lobby");
            controller.updateWaitingPlayers(players);
        });

    }

    public void joinPlayer(String oppId) {
        Request request = new Request(RequestType.JOIN_PLAYER);
        request.setData(oppId);
        ClientService.getInstance().sendRequest(request);
    }
}

File: game-client/src/main/kotlin/io/github/octcarp/sustech/cs209a/linkgame/client/net/LobbyData.kt
package io.github.octcarp.sustech.cs209a.linkgame.client.net

import io.github.octcarp.sustech.cs209a.linkgame.client.controller.LobbyController
import io.github.octcarp.sustech.cs209a.linkgame.client.controller.MainMenuController
import io.github.octcarp.sustech.cs209a.linkgame.client.utils.SceneSwitcher
import io.github.octcarp.sustech.cs209a.linkgame.common.packet.Request
import io.github.octcarp.sustech.cs209a.linkgame.common.packet.RequestType

object LobbyData {
    private var isWaiting = false

    fun enterLobby() = ClientService.sendRequestOnlyType(RequestType.ENTER_LOBBY)

    fun exitLobby() = ClientService.sendRequestOnlyType(RequestType.EXIT_LOBBY)

    fun toggleWaiting() {
        ClientService.sendRequestOnlyType(if (isWaiting) RequestType.STOP_WAITING else RequestType.START_WAITING)
        isWaiting = !isWaiting
    }

    fun reconnectMatch() = ClientService.sendRequestOnlyType(RequestType.RECONNECT_MATCH)

    fun reAllWaitingPlayers(players: List<String>) {
        (SceneSwitcher.getController("lobby") as LobbyController).updateWaitingPlayers(players)
    }

    fun joinPlayer(oppId: String) {
        Request(RequestType.JOIN_PLAYER).apply {
            data = oppId
        }.also(ClientService::sendRequest)
    }

    fun reNoReconnect() {
        (SceneSwitcher.getController("main-menu") as MainMenuController).handleNoReconnect()
    }
}
